### ISSUE TITLE
IncompatibleClassChangeError when implementing Java annotation

### ISSUE THREAD

**Issue by cchantep** [lampepfl/dotty#14199]
See that the previous issue is back: https://github.com/scala/bug/issues/8778

## Compiler version

3.1.2-RC1-bin-20211222-c94b333-NIGHTLY

## Minimized code

```java
import javax.inject.Qualifier;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

@Qualifier
@Retention(RetentionPolicy.RUNTIME)
public @interface NamedDatabase {
    String value();
}
```

```scala
import java.io.Serializable

final class NamedDatabaseImpl(val value: String)
    extends NamedDatabase(value)
    with Serializable {

  assert(value != null)

  override def hashCode: Int = (127 * "value".hashCode()) ^ value.hashCode()

  override def equals(that: Any): Boolean = that match {
    case other: NamedDatabase =>
      other.value == this.value

    case _ =>
      false
  }

  override def toString: String =
    s"@${classOf[NamedDatabase].getName}(value=${value})"

  def annotationType = classOf[NamedDatabase]
}
```

## Output

When try to use it with Guice, fails with a `IncompatibleClassChangeError`

```
[error]  java.lang.IncompatibleClassChangeError: play/modules/reactivemongo/NamedDatabaseImpl (ReactiveMongoModule.scala:48
[error] play.modules.reactivemongo.ReactiveMongoModule$.key(ReactiveMongoModule.scala:48)
[error] play.modules.reactivemongo.ReactiveMongoModule.apiBindings$$anonfun$1(ReactiveMongoModule.scala:36)
[error] play.modules.reactivemongo.ReactiveMongoModule$$Lambda$12039/1390106650.apply(Unknown Source)
[error] play.modules.reactivemongo.ReactiveMongoModule.apiBindings(ReactiveMongoModule.scala:41)
[error] play.modules.reactivemongo.ReactiveMongoModule.bindings(ReactiveMongoModule.scala:24)
[error] play.modules.reactivemongo.ReactiveMongoModule.bindings(ReactiveMongoModule.scala:17)
[error] play.api.inject.guice.GuiceableModuleConversions.guice(GuiceInjectorBuilder.scala:365)
[error] play.api.inject.guice.GuiceableModuleConversions.guice$(GuiceInjectorBuilder.scala:364)
[error] play.api.inject.guice.GuiceableModule$.guice(GuiceInjectorBuilder.scala:297)
```

**Comment by smarter** [comment]
Same root cause as https://github.com/lampepfl/dotty/issues/12840#issuecomment-862339051, so no easy fix.

**Comment by cchantep** [comment]
That's quite a regression considering the fix applied about that in 2.13, no easy workaround.

**Comment by smarter** [comment]
Yep, that's what happens when two compilers are developed in parallel.
